src/main/java/hudson/plugins/build_timeout/BuildTimeoutWrapper.java

```diff
@@ -193,6 +195,8 @@ public synchronized void rescheduleIfScheduled() {
             public synchronized boolean tearDown(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
                 if (task != null) {
                     task.cancel();
+                    // avoid memory leaks for the case where this timer is in the future (JENKINS-31627).
+                    Trigger.timer.purge();
```

Comment:
@tfennelly 
Task is a [`java.util.TimerTask`](http://docs.oracle.com/javase/7/docs/api/index.html?java/util/TimerTask.html) that is in a [`java.util.Timer`](http://docs.oracle.com/javase/7/docs/api/index.html?java/util/Timer.html)

You cancel the `task` so that when the `timer` comes to run it it won't run it - but the `timer` still holds a reference to it.